pkg/updater/repository/repository.go

```diff
@@ -204,34 +236,49 @@ func readRepository(rootPath string) (*repositoryFiles, error) {
 
 	return &repositoryFiles{
 		rootPath:   rootPath,
+		runPath:    runPath,
 		stable:     stableLink,
 		experiment: experimentLink,
 	}, nil
 }
 
 func (r *repositoryFiles) setExperiment(name string, sourcePath string) error {
-	path, err := movePackageFromSource(name, r.rootPath, sourcePath)
+	path, err := movePackageFromSource(name, r.rootPath, r.runPath, sourcePath)
 	if err != nil {
 		return fmt.Errorf("could not move experiment source: %w", err)
 	}
+
 	return r.experiment.Set(path)
 }
 
 func (r *repositoryFiles) setStable(name string, sourcePath string) error {
-	path, err := movePackageFromSource(name, r.rootPath, sourcePath)
+	path, err := movePackageFromSource(name, r.rootPath, r.runPath, sourcePath)
 	if err != nil {
 		return fmt.Errorf("could not move stable source: %w", err)
 	}
+
 	return r.stable.Set(path)
 }
 
-func movePackageFromSource(packageName string, rootPath string, sourcePath string) (string, error) {
+func movePackageFromSource(packageName string, rootPath string, runPath string, sourcePath string) (string, error) {
 	if packageName == "" || packageName == stableVersionLink || packageName == experimentVersionLink {
 		return "", fmt.Errorf("invalid package name")
 	}
 	targetPath := filepath.Join(rootPath, packageName)
 	_, err := os.Stat(targetPath)
 	if err == nil {
+		// Check if we have long running processes using the package
+		// If yes, the GC left the package in place so we don't reinstall it, but
+		// we don't throw an error either.
+		// If not, the GC should have removed the packages so we error.
+		targetRunPath := filepath.Join(runPath, packageName)
+		inUse, err := versionUsed(targetRunPath)
```

Comment:
This pattern of calling `versionUsed` twice (once during cleanup and once during this move) looks dangerous to me. If the injector or something else was to delete the PID file on disk you would have an unexpected error here for example.

I think it's much better to load the list of locked packages in `readRepository` and use that cached state during the full operation. It will also reduce the number of error checks you have to do. LMK if you want to chat about this.

Reply:
Done in [58e16d5](https://github.com/DataDog/datadog-agent/pull/22243/commits/58e16d5897e1a2a6ae0bd6fcf343579078c99b9e)